File: Frontend/Swift/barbicha/Model/AppointmentType.swift
//
//  AppointmentType.swift
//  Barbicha
//
//  Created by Gustavo Tavares on 17/05/2018.
//  Copyright © 2018 Gustavo Tavares. All rights reserved.
//

import Foundation
import SipHash

struct AppointmentType: Equatable, Codable {
    
    let index: Int
    let label: String
    let minutes: Int
    
    var time: TimeInterval {return Double(self.minutes) * 60}
    
    static let kEmptyTypeIndex:Int = -1
    static let kUnavailableTypeIndex: Int = -2
    
    static var empty: AppointmentType {return AppointmentType(index: AppointmentType.kEmptyTypeIndex, label: "", minutes: 30)}
    static var unavailable: AppointmentType {return AppointmentType(index: AppointmentType.kUnavailableTypeIndex, label: "", minutes: 30)}
    
}

extension AppointmentType: CustomStringConvertible {
    
    var description: String {
        return "AppointmentType [\(self.index) - \(self.label), \(self.time) ]"
    }
    
    
}

extension AppointmentType: SipHashable {
    
    func appendHashes(to hasher: inout SipHasher) {
        hasher.append(self.index)
        hasher.append(self.label)
        hasher.append(self.minutes)
    }

}

File: Frontend/Android/app/src/main/java/br/com/plazaz/barbicha/model/AppointmentType.kt
package br.com.plazaz.barbicha.model

data class AppointmentType(val index:Int, val label:String, val minutes: Int) {

    companion object {

        const val kIndex = "index"
        const val kLabel = "label"
        const val kTime = "minutes"


        const val kEmptyTypeIndex:Int = -1
        const val kUnavailableTypeIndex: Int = -2

        fun empty(): AppointmentType {return AppointmentType(kEmptyTypeIndex, "", 30)}
        fun unavailable(): AppointmentType {return AppointmentType(kUnavailableTypeIndex, "", 30)}

        fun fromMap(map: Map<String, Any>): AppointmentType {

            val index = map.get(kIndex) as Long
            val label = map.get(kLabel) as String
            val minutes = map.get(kTime) as Long
            return AppointmentType(index.toInt() , label, minutes.toInt())

        }

    }

}